core/src/services/huggingface/core.rs

```diff
@@ -0,0 +1,415 @@
+
+        if let Some(token) = &self.token {
+            let auth_header_content = format_authorization_by_bearer(token)?;
+            req = req.header(header::AUTHORIZATION, auth_header_content);
+        }
+
+        req = req.header(header::CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+        let req_body = format!("paths={}&expand=True", percent_encode_path(&p));
+
+        let req = req
+            .body(AsyncBody::Bytes(Bytes::from(req_body)))
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn hf_list(
+        &self,
+        path: &str,
+        recursive: bool,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path)
+            .trim_end_matches('/')
+            .to_string();
+
+        let mut url = match self.repo_type {
+            RepoType::Model => format!(
+                "https://huggingface.co/api/models/{}/tree/{}/{}?expand=True",
+                &self.repo_id,
+                &self.revision,
+                percent_encode_path(&p)
+            ),
+            RepoType::Dataset => format!(
+                "https://huggingface.co/api/datasets/{}/tree/{}/{}?expand=True",
+                &self.repo_id,
+                &self.revision,
+                percent_encode_path(&p)
+            ),
+        };
+
+        if recursive {
+            url.push_str("&recursive=True");
+        }
+
+        let mut req = Request::get(&url);
+
+        if let Some(token) = &self.token {
+            let auth_header_content = format_authorization_by_bearer(token)?;
+            req = req.header(header::AUTHORIZATION, auth_header_content);
+        }
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn hf_read(&self, path: &str, arg: OpRead) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path)
+            .trim_end_matches('/')
+            .to_string();
+
+        let url = match self.repo_type {
+            RepoType::Model => format!(
+                "https://huggingface.co/{}/resolve/{}/{}",
+                &self.repo_id,
+                &self.revision,
+                percent_encode_path(&p)
+            ),
+            RepoType::Dataset => format!(
+                "https://huggingface.co/datasets/{}/resolve/{}/{}",
+                &self.repo_id,
+                &self.revision,
+                percent_encode_path(&p)
+            ),
+        };
+
+        let mut req = Request::get(&url);
+
+        if let Some(token) = &self.token {
+            let auth_header_content = format_authorization_by_bearer(token)?;
+            req = req.header(header::AUTHORIZATION, auth_header_content);
+        }
+
+        let range = arg.range();
+        if !range.is_full() {
+            req = req.header("Range", &range.to_header());
```

Comment:
Please always use `http::header` for well-known header names.

Reply:
Done